lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -2770,6 +2771,338 @@ def permutation_matrices(n):
+            **base_style,
+            **kwargs,
+            'zorder': (kwargs['zorder'] - .1 if barsabove else
+                       kwargs['zorder'] + .1),
+        }
+
+        # make the style dict for the line collections (the bars)
+        eb_lines_style = dict(base_style)
+        eb_lines_style.pop('marker', None)
+        eb_lines_style.pop('markerfacecolor', None)
+        eb_lines_style.pop('markeredgewidth', None)
+        eb_lines_style.pop('markeredgecolor', None)
+        eb_lines_style.pop('linestyle', None)
+        eb_lines_style['color'] = ecolor
+
+        if elinewidth:
+            eb_lines_style['linewidth'] = elinewidth
+        elif 'linewidth' in kwargs:
+            eb_lines_style['linewidth'] = kwargs['linewidth']
+
+        for key in ('transform', 'alpha', 'zorder', 'rasterized'):
+            if key in kwargs:
+                eb_lines_style[key] = kwargs[key]
+
+        # make the style dict for cap collections (the "hats")
+        eb_cap_style = dict(base_style)
+        # eject any marker information from format string
+        eb_cap_style.pop('marker', None)
+        eb_cap_style.pop('ls', None)
+        eb_cap_style['linestyle'] = 'none'
+        if capsize is None:
+            capsize = kwargs.pop('capsize', rcParams["errorbar.capsize"])
+        if capsize > 0:
+            eb_cap_style['markersize'] = 2. * capsize
+        if capthick is not None:
+            eb_cap_style['markeredgewidth'] = capthick
+        eb_cap_style['color'] = ecolor
+
+        if plot_line:
+            data_line = art3d.Line3D(x, y, z, **plot_line_style)
+            self.add_line(data_line)
+
+        try:
+            offset, errorevery = errorevery
+        except TypeError:
+            offset = 0
+
+        if errorevery < 1 or int(errorevery) != errorevery:
+            raise ValueError(
+                'errorevery must be positive integer or tuple of integers')
+        if int(offset) != offset:
+            raise ValueError("errorevery's starting index must be an integer")
+
+        everymask = np.zeros(len(x), bool)
+        everymask[offset::errorevery] = True
+
+        def _mask_lists(xs, ys, zs, mask=None):
+            """Applies a mask to three lists."""
+            xs = [l for l, m in zip(xs, mask) if m]
+            ys = [l for l, m in zip(ys, mask) if m]
+            zs = [l for l, m in zip(zs, mask) if m]
+            return xs, ys, zs
+
+        # TODO: errors can be only a scalar number or len(N) array-like
+        def _unpack_errs(err, data, lomask, himask):
+            # for compatibility with the 2d errorbar function, when both upper
+            # and lower limits specified, we need to draw the markers / line -
+            # whether or not using both limits makes any sense (it doesn't)
+            _lomask = lomask | ((lomask == himask) & everymask)
+            _himask = himask | ((lomask == himask) & everymask)
+
+            lows = [d - e if m else d for d, e, m in zip(data, err, _lomask)]
```

Comment:
```suggestion
            lows = np where(_lomask, data - err, data)
```

Reply:
@eric-wieser the 2d function I've adapted this from makes a big fuss about not using arrays for data (e.g. `xywhere`, the [equivalent of this function](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axes/_axes.py#L3372-L3381)), so I tried to use numpy arrays for masking only. If we can ditch that and just use arrays, it would simplify things a lot, but I'm not sure that's the case.

Reply:
That's good to know, thanks. I know a lot of work recently went into trying to make `mplot3d` faster by using arrays more, so it would be good to know why the 2d one avoids it.

Reply:
Now that I look at it, there's a FIXME item down the file where I've actually converted everything into arrays. I'll add fixing that to my todo-list for the PR.

Reply:
Meanwhile, if anyone knows of a testable usecase where using arrays is not ok, I'd like to see it (for testing reasons).

Reply:
@eric-wieser I don't know anymore why the 2d version does this the way it does.  The original code comes from pre-2007 commits (as far as git blame took me), and I don't even know those comments are relevant, because
1. I forced np.array on errors in 2d errorbar and ran `pytest lib/matplotlib/tests/test_axes.py` and `pytest lib/matplotlib/tests/test_units.py` successfully,
2. the only other place I've found units being an issue (https://github.com/matplotlib/matplotlib/pull/10118) was closed with only upstream changes, and
3. Some special cases, such as plotting pandas dataframes, appear to work normally regardless of forcing arrays

Given that I'm already using arrays for errors elsewhere in the code, can I suggest I implement the change here and the unit business gets resolved in a separate issue (as I don't seem to understand it)?